### ISSUE TITLE
RxCocoa +load conflicts at link time in Xcode 12

### ISSUE THREAD

**Issue by jshier** [ReactiveX/RxSwift#2195]
**Short description of the issue**:

Xcode 12 beta 1 produces a compile warning when linking: "ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category"

**Expected outcome**:

No warnings.

**What actually happens**:

Repeated linker warnings:

```
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
ld: warning: method '+load' in category from ...RxCocoa.build/Objects-normal-tsan/x86_64/_RXObjCRuntime.o conflicts with same method from another category
```

**Self contained code example that reproduces the issue**:

N/AA

**RxSwift/RxCocoa/RxBlocking/RxTest version/commit**

RxCocoa 5.1.1

**Platform/Environment**

  - [x] iOS

**How easy is to reproduce? (chances of successful reproduce after running the self contained code)**

  - [x] easy, 100% repro

**Xcode version**:

```
Version 12.0 beta (12A6159)
```

**Installation method**:
  - [x] CocoaPods
  - [ ] Carthage
  - [ ] Git submodules

**I have multiple versions of Xcode installed**:
  (so we can know if this is a potential cause of your issue)
  - [x] yes (which ones) 12b1, 11.5, 11.4, 11.3, 10.2
  - [ ] no

**Level of RxSwift knowledge**:
  (this is so we can understand your level of knowledge
    and formulate the response in an appropriate manner)
  - [ ] just starting
  - [x] I have a small code base
  - [ ] I have a significant code base

**Comment by freak4pc** [comment]
Hey @jshier, thanks for reporting this :)
We usually don't release fixes on early betas, but I'm gonna tag this so we can keep track of it.
If it's still an issue around Xcode 12 beta 4, we'll apply a fix.

Thanks! 
Shai

**Comment by alex-vasenin** [comment]
I have no idea if this warning is relevant or not, but If you want to just _disable_ it, you can do so in your post install hook:
```
post_install do |installer|
  installer.pods_project.targets.each do |target|
    if target.name == "RxCocoa"
      target.build_configurations.each do |config|
        config.build_settings["OTHER_LDFLAGS"] = "-w"
      end
    end
  end
end
```

**Comment by scottasoutherland** [comment]
This still exists as of Beta 2 (12A6163b)

**Comment by freak4pc** [comment]
Thanks, Scott :) 
This seems quite weird, but I'll see if I can take a look soon.

**Comment by sharplet** [comment]
I did some digging on this, inspecting the preprocessed Objective-C code, the symbols defined by RxCocoaRuntime.o and the resulting RxCocoa binary. I struggled to find any evidence of duplicated category methods.

I also performed the following experiment:

1. Build RxCocoa in Xcode 12 beta.
2. Copy the linker invocation that produces the warnings.
3. Without changing anything else, change /Applications/Xcode-beta.app to /Applications/Xcode.app in the path to the linker command (i.e., `clang`).
4. Observe whether Xcode 11's linker produces the same warnings when given exactly the same intermediate object files as input.

The answer? Xcode 11 gives no warnings, which is the strongest evidence I could find that maybe this warning is spurious and could be a bug with the Xcode 12 linker.

**Comment by sharplet** [comment]
I've filed FB7982824 to track whether this is a linker bug.

**Comment by cltnschlosser** [comment]
This warning makes sense though right? SWIZZLE_OBSERVE_METHOD macro is called 27 times. There are 26 warnings. What the macro does is make a new category on RXObjCRuntime with a unique name and each category has a `+load`. So on the same object you get 27 categories all with the same method in them. Therefore 26 duplicates = 26 warnings.

I think what would need to be done is either make a new object everytime (kind of a waste), or make one `+load` that has the body of all the previous 27. That shouldn't be that complex of a refactor.

**Comment by freak4pc** [comment]
@cltnschlosser Happy to accept a PR on your latter offer if you have thoughts. Make sure you target the `rxswift6` branch, if you do, since that would be the RC for Xcode 12 :)

**Comment by jshier** [comment]
Xcode 12b3 is still producing this warning, FYI.

**Comment by freak4pc** [comment]
That makes sense, thank you!

**Comment by sharplet** [comment]
I couldn't find a canonical reference for this, but from inspecting the RxCocoa binary produced by Xcode 11, there were clearly 27 `+load` methods defined, and in a small test project the dynamic loader definitely treated them as distinct entry points. What I'm not sure about is whether this behaviour is by design or essentially UB. Put differently: is Xcode 12 producing a warning that it should have been producing all along, or is this a supported use case and the warning is spurious?

**Comment by freak4pc** [comment]
That's the same question I'm asking myself. Regardless, I'll try to work around this based on Colton's suggestion but I had really hoped not to mess with that area due to how sensitive it is.

**Comment by freak4pc** [comment]
It definitely seems like _some sort_ of a regression since even the unit tests that were aimed at making sure this are failing now

![image](https://user-images.githubusercontent.com/605076/88228261-95864280-cc77-11ea-805e-5b1aa11deef8.png)

**Comment by cltnschlosser** [comment]
I think I have a fix. Can have a PR up shortly.

**Comment by freak4pc** [comment]
That would be perfect. Thank you!

**Comment by freak4pc** [comment]
Fixed in #2209. Thanks @cltnschlosser for your awesome help!
